pkg/controllers/management/auth/project_cluster/cluster_handler.go

```diff
@@ -319,66 +332,69 @@ func (l *clusterLifecycle) addRTAnnotation(obj runtime.Object, context string) (
+		cluster, ok := obj.(*apisv3.Cluster)
+		if !ok {
+			return obj, fmt.Errorf("expected cluster, got %T", obj)
 		}
 
-		typeAccessor, err := meta.TypeAccessor(obj)
-		if err != nil {
-			return obj, err
-		}
-
-		creatorID, ok := metaAccessor.GetAnnotations()[CreatorIDAnnotation]
-		if !ok || creatorID == "" {
-			logrus.Warnf("[%v] %v %v has no creatorId annotation. Cannot add creator as owner", ClusterCreateController, typeAccessor.GetKind(), metaAccessor.GetName())
+		creatorID := cluster.Annotations[CreatorIDAnnotation]
+		if creatorID == "" {
+			logrus.Warnf("[%s] cluster %s has no creatorId annotation. Cannot add creator as owner", ClusterCreateController, cluster.Name)
 			return obj, nil
 		}
-		cluster := obj.(*apisv3.Cluster)
 
 		if apisv3.ClusterConditionInitialRolesPopulated.IsTrue(cluster) {
 			// The clusterRoleBindings are already completed, no need to check
 			return obj, nil
 		}
 
-		roleJSON := cluster.Annotations[roleTemplatesRequiredAnnotation]
-		if roleJSON == "" {
+		creatorRoleBindings := cluster.Annotations[roleTemplatesRequiredAnnotation]
+		if creatorRoleBindings == "" {
 			return cluster, nil
 		}
 
 		roleMap := make(map[string][]string)
-		err = json.Unmarshal([]byte(roleJSON), &roleMap)
+		err := json.Unmarshal([]byte(creatorRoleBindings), &roleMap)
 		if err != nil {
 			return obj, err
 		}
 
 		var createdRoles []string
-
 		for _, role := range roleMap["required"] {
 			rtbName := "creator-" + role
 
-			if rtb, _ := l.crtbLister.Get(metaAccessor.GetName(), rtbName); rtb != nil {
+			if rtb, _ := l.crtbLister.Get(cluster.Name, rtbName); rtb != nil {
 				createdRoles = append(createdRoles, role)
 				// This clusterRoleBinding exists, need to check all of them so keep going
 				continue
 			}
 
 			// The clusterRoleBinding doesn't exist yet so create it
-			om := metav1.ObjectMeta{
-				Name:      rtbName,
-				Namespace: metaAccessor.GetName(),
-			}
-			om.Annotations = crtbCreatorOwnerAnnotations
-
-			logrus.Infof("[%v] Creating creator clusterRoleTemplateBinding for user %v for cluster %v", ClusterCreateController, creatorID, metaAccessor.GetName())
-			if _, err := l.mgr.Management.ClusterRoleTemplateBindings(metaAccessor.GetName()).Create(&apisv3.ClusterRoleTemplateBinding{
-				ObjectMeta:       om,
-				ClusterName:      metaAccessor.GetName(),
+			crtb := &apisv3.ClusterRoleTemplateBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:        rtbName,
+					Namespace:   cluster.Name,
+					Annotations: crtbCreatorOwnerAnnotations,
+				},
+				ClusterName:      cluster.Name,
 				RoleTemplateName: role,
 				UserName:         creatorID,
-			}); err != nil && !apierrors.IsAlreadyExists(err) {
+			}
+
+			if principalName := cluster.Annotations[creatorPrincipalNameAnnotation]; principalName != "" {
+				if !strings.HasPrefix(principalName, "local") {
+					// Setting UserPrincipalName only makes sense for non-local users.
+					crtb.UserPrincipalName = principalName
+					crtb.UserName = ""
```

Comment:
why do we need to set `UserName` to be empty here? I might be missing something, but I think this might prevent `CRTBs` from being synced here https://github.com/rancher/rancher/blob/v2.9.1/pkg/controllers/managementuser/rbac/crtb_handler.go#L62

Reply:
oh, good point.  That does seem to be the case.

Reply:
That's a great question, thanks for picking up on that!
> I think this might prevent CRTBs from being synced here

It's counterintuitive but it works correctly as `Username` is reinstated by `reconcileSubject` call. What's even more important this ensures that the principal belongs to that user.

Technically we can avoid this additional work (as the name is known ahead of time) by setting both and relying 💯 on the webhook validation.

Reply:
BTW it's exactly what happens when you add a Member to an existing cluster/project. In case you'd have a payload that looks along the lines of
```json
POST https://<rancher-url>/v3/projectroletemplatebindings
{
    "type": "projectroletemplatebinding",
    "roleTemplateId": "project-member",
    "userPrincipalId": "keycloak_user://testuser2",
    "projectId": "c-t9drq:p-qwl7j"
}
```

Reply:
Thanks for the clarification!